presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java

```diff
@@ -0,0 +1,305 @@
+                case SYNC_FAILURE:
+                    throw new SyncException("Failed on call " + taskTracker.totalAttemptsMade());
+                case ASYNC_FAILURE:
+                    CompletableFuture<Integer> result = new CompletableFuture<>();
+                    result.completeExceptionally(new AsyncException("Failed on call " + taskTracker.totalAttemptsMade()));
+                    return toListenableFuture(result);
+                default:
+                    throw new RuntimeException("Unexpected exception type");
+            }
+        }
+    }
+
+    @Test
+    void testNoRetry()
+            throws Exception
+    {
+        SyncTaskImpl task = new SyncTaskImpl(SUCCESS);
+        Integer result = retry.run("test", task);
+
+        assertSyncSuccess(result, task, 1);
+    }
+
+    @Test
+    void testRetrySuccess()
+            throws Exception
+    {
+        SyncTaskImpl task = new SyncTaskImpl(SYNC_FAILURE, SYNC_FAILURE, SYNC_FAILURE, SYNC_FAILURE, SUCCESS);
+        Integer result = retry.run("test", task);
+
+        assertSyncSuccess(result, task, 5);
+    }
+
+    @Test
+    void testOutOfRetries()
+    {
+        SyncTaskImpl task = new SyncTaskImpl(SYNC_FAILURE, SYNC_FAILURE, SYNC_FAILURE);
+
+        try {
+            retry.maxAttempts(3).run("test", task);
+            fail("Call didn't fail as expected");
+        }
+        catch (Exception e) {
+            if (!(e instanceof SyncException)) {
+                fail("Expected " + SyncException.class + ", but got " + e.getClass());
+            }
+        }
+        assertEquals(task.totalAttemptsMade(), 3);
+    }
+
+    @Test
+    void testAsyncNoRetry()
+            throws Exception
+    {
+        AsyncTaskImpl task = new AsyncTaskImpl(SUCCESS);
+        ListenableFuture<Integer> result = retry.runAsync("test", task);
+
+        assertAsyncSuccess(result, task, 1);
+    }
+
+    @Test
+    void testAsyncMixedRetries()
+            throws Exception
+    {
+        AsyncTaskImpl task = new AsyncTaskImpl(SYNC_FAILURE, ASYNC_FAILURE, SYNC_FAILURE, ASYNC_FAILURE, SUCCESS);
+        ListenableFuture<Integer> result = retry.runAsync("test", task);
+
+        assertAsyncSuccess(result, task, 5);
+    }
+
+    @Test
+    void testAsyncMultipleAsyncFailures()
+            throws Exception
+    {
+        AsyncTaskImpl task = new AsyncTaskImpl(ASYNC_FAILURE, ASYNC_FAILURE, ASYNC_FAILURE, ASYNC_FAILURE, SUCCESS);
+        ListenableFuture<Integer> result = retry.runAsync("test", task);
+
+        assertAsyncSuccess(result, task, 5);
+    }
+
+    @Test
+    void testAsyncOutOfRetries()
+            throws Exception
+    {
+        AsyncTaskImpl task = new AsyncTaskImpl(ASYNC_FAILURE, ASYNC_FAILURE, ASYNC_FAILURE, ASYNC_FAILURE);
+        ListenableFuture<Integer> result = retry.maxAttempts(4).runAsync("test", task);
+
+        assertAsyncExceptionally(result, task, AsyncException.class, 4);
+    }
+
+    @Test
+    void testSyncOutOfRetries()
```

Comment:
this test is the same as `testOutOfRetries`

Reply:
Yes they are the same internally. Will remove one of them.